### PR TITLE
Add C code generation backend

### PR DESCRIPTION
**Pull request by paulcadman**
This PR adds a C code generation backend (available in the `minijuvix minic` command).

## Unsupported

The C backend does not currently support:
* higher-order functions
* currying
* mutually recursive functions

## Try it out

```shell
$ cd tests/positive/MiniC/HelloWorld
$ minijuvix minic Input.mjuvix | emcc -x c - -o out.wasm && wasmer out.wasm
hello world!
```

```shell
$ cd tests/positive/MiniC/Nat
$ minijuvix minic Input.mjuvix | emcc -x c - -o out.wasm && wasmer out.wasm
3 + (1 + 1) = 5
3 == 1 + 2 : True
3 == 2 : False
is-even 3 : False
is-even 4 : True
```

## Tests

The tests now depend on [emscripten](https://emscripten.org) and [wasmer](https://wasmer.io).  The GitHub action has been updated to support this.

I've tested this with Emscripten version 3.1.8 and Wasmer 2.2.1. The CI runs tests against the latest version of each.

You can skip running the C->WASM generation and execution tests by running:

```
stack test --fast --ta '-p "! /slow tests/"'
```

or

```
make test-skip-slow 
```

closes https://github.com/heliaxdev/minijuvix/issues/60

### CODE REVIEW

File: app/Commands/MiniC.hs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{-# LANGUAGE ApplicativeDo #-}
+
+module Commands.MiniC where
+
+import Commands.Extra
+import MiniJuvix.Prelude hiding (Doc)
+import Options.Applicative
+
+newtype MiniCOptions = MiniCOptions
+  { _mcInputFile :: FilePath
```
Comment (jonaprieto) [inline app/Commands/MiniC.hs:10]: For mcInputFile, either → miniCInputFile or miniCOptionsInputFile

File: src/MiniJuvix/Syntax/Backends.hs [hunk 2]
```diff
@@ -2,7 +2,7 @@ module MiniJuvix.Syntax.Backends where
 
 import MiniJuvix.Prelude
 
-data Backend = BackendGhc | BackendAgda
+data Backend = BackendGhc | BackendAgda | BackendC
```
[the remark this answers is not in the document]
Comment (paulcadman) [reply src/MiniJuvix/Syntax/Backends.hs:5]: BackendAgda is in dev - https://github.com/heliaxdev/minijuvix/blob/7b1371c4b95752c23b54d40fe781f356c0ffb8dc/src/MiniJuvix/Syntax/Backends.hs#L5.

So I think we should remove it in a separate PR.

[the remark this answers is not in the document]
Comment (paulcadman) [reply src/MiniJuvix/Syntax/Backends.hs:5]: BackendAgda is in dev - https://github.com/heliaxdev/minijuvix/blob/7b1371c4b95752c23b54d40fe781f356c0ffb8dc/src/MiniJuvix/Syntax/Backends.hs#L5.

So I think we should remove it in a separate PR.

File: package.yaml [hunk 3]
```diff
@@ -31,13 +31,15 @@ dependencies:
...
 - hashable == 1.4.*
+- language-c == 0.9.*
 - megaparsec == 9.2.*
 - microlens-platform == 0.4.*
 - parser-combinators == 1.3.*
 - polysemy == 1.7.*
 - polysemy-plugin == 0.4.*
 - prettyprinter == 1.7.*
 - prettyprinter-ansi-terminal == 1.1.*
+- pretty == 1.1.3.*
```
Comment (jonaprieto) [inline package.yaml:42]: Why is this really needed?

Comment (paulcadman) [reply package.yaml:42]: Yes - it is the pretty printer used by the language-c package that is used by the C code serialiser: https://github.com/heliaxdev/minijuvix/blob/ee6eac9e4992c83ddc80e55c65ab5605966db858/src/MiniJuvix/Syntax/CJuvix/Serialization.hs#L9

**Review by janmasrovira — CHANGES_REQUESTED** [review]
C review

File: app/Main.hs [hunk 4]
```diff
@@ -32,6 +33,7 @@ import MiniJuvix.Termination qualified as Termination
 import MiniJuvix.Termination.CallGraph qualified as Termination
```
Comment (jonaprieto) [inline app/Main.hs:33]: Can you please remove this line? It is redundant.

**Review by jonaprieto — DISMISSED** [review]

**Review by janmasrovira — APPROVED** [review]

**Pull request merged**